File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        pivot = 0
        r = len(nums) - 1
        l = pivot + 1

        while (pivot < len(nums) - 2):
            if pivot > 0 and nums[pivot] == nums[pivot - 1]:
                pivot += 1
                l = pivot + 1
                r = len(nums) - 1
                continue
            while (l < r):
                if nums[pivot] + nums[l] + nums[r] > 0:
                    r -= 1
                elif nums[pivot] + nums[l] + nums[r] < 0:
                    l += 1
                else:
                    res.append([nums[pivot], nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1 
            pivot += 1
            l = pivot + 1
            r = len(nums) - 1

        return res

File: Data Structures & Algorithms/reconstruct-flight-path/submission-0.cpp
class Solution {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        std::unordered_map<std::string, std::deque<std::string>> adj;
        for (auto t : tickets) {
            adj[t[0]].push_back(t[1]);
        }
        for (auto& [src, dests] : adj) {
            std::sort(dests.rbegin(), dests.rend());
        }

        std::vector<std::string> res;
        res.reserve(tickets.size() + 1);

        dfs("JFK", adj, res);
        std::reverse(res.begin(), res.end());

        return res;


    }

    void dfs(const std::string& city, std::unordered_map<std::string, std::deque<std::string>>& adj, std::vector<std::string>& res) {

        while(!adj[city].empty()) {
            std::string dest = adj[city].back();
            adj[city].pop_back();
            dfs(dest, adj, res);
        }

        res.push_back(city);
    }

    
};
